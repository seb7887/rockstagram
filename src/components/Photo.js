import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import LikeAnimation from './LikeAnimation';

const Figure = styled.figure`
  flex-basis: calc(33.33% - 4rem);
  flex-grow: 1;
  flex-shrink: 0;
  margin: 0 2rem 2rem 2rem;
  padding: 2rem;
  border: 1px solid var(--lightgrey);
  background: white;
  box-shadow: 0 0 0 3.5px rgba(0, 0, 0, 0.01);
  position: relative;

  .grid-photo-wrap {
    position: relative;
  }

  .grid-photo {
    width: calc(100% + 4rem);
    margin-left: -2rem;
    margin-top: -2rem;
    max-width: none;
  }

  .control-buttons {
    display: flex;
    justify-content: space-between;
    width: 100%;
  }

  .button {
    border: 2px solid rgba(0, 0, 0, 0.2);
    background: none;
    width: 100%;
    display: inline-block;
    line-height: 2;
    text-decoration: none;
    padding: 1rem 4rem;
    text-align: center;
    font-size: 2.5rem;
    color: var(--black);
    transition: all 0.2s;
    box-sizing: padding-box;


    &:hover, &:focus {
      border-color: var(--black);
      opacity: 0.8;
      outline: 0;
    }
  }

  .likes {
    padding: 1rem 4rem;
    font-size: 2.5rem;
    width: 100%;
    background: transparent;
    border: 2px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    transition: all 0.2s;

    &:hover, &:focus {
      border-color: var(--black);
      opacity: 0.8;
      outline: 0;
    }
  }

  .comment-count {
    font-size: 2.5rem;
    
    i {
      vertical-align: middle;
    }
  }

  .speech-bubble {
    width: 1.25rem * 1.2
    height: 1.25rem;
    background: var(--black);
    display: inline-block;
    border-radius: 50%;
    position: relative;
  }
`;

const Photo = ({ post, i, comments, incrementLikes }) => (
  <Figure>
    <div className='grid-photo-wrap'>
      <Link to={`/view/${post.code}`}>
        <img src={post.display_src} alt={post.caption} className='grid-photo' />
      </Link>

      <LikeAnimation likes={post.likes} />
    </div>

    <figcaption>
      <p>{post.caption}</p>
      <div className='control-buttons'>
        <button className='likes' onClick={incrementLikes.bind(null, i)}><i className='fa fa-heart' /> {post.likes}</button>
        <Link to={`/view/${post.code}`} className='button'>
          <span className='comment-count'>
            <i className='fa fa-comment' /> {comments[post.code] ? comments[post.code].length : 0}
          </span>
        </Link>
      </div>
    </figcaption>
  </Figure>
)

export default Photo;
