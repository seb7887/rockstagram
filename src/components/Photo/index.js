import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { commentsSelector } from '../../store/selectors';
import { incrementLikes } from '../../store/actions/actionCreators';
import { Figure } from './style';

import LikeAnimation from '../LikeAnimation';

const Photo = ({ post, i, comments, incrementLikes }) => (
  <Figure>
    <div className='grid-photo-wrap'>
      <Link to={`/view/${post.code}`}>
        <img src={post.display_src} alt={post.caption} className='grid-photo' />
      </Link>

      <LikeAnimation likes={post.likes} />
    </div>

    <figcaption data-testid='caption'>
      <p>{post.caption}</p>
      <div className='control-buttons'>
        <button className='likes' onClick={incrementLikes.bind(null, i)}>
          <i className='fa fa-heart' data-testid='likes' /> {post.likes}
        </button>
        <Link to={`/rockstagram/view/${post.code}`} className='button'>
          <span className='comment-count' data-testid='count'>
            <i className='fa fa-comment' />{' '}
            {comments[post.code] ? comments[post.code].length : 0}
          </span>
        </Link>
      </div>
    </figcaption>
  </Figure>
);

const mapStateToProps = state => {
  const comments = commentsSelector(state);

  return {
    comments,
  };
};

export default connect(
  mapStateToProps,
  { incrementLikes },
)(Photo);
