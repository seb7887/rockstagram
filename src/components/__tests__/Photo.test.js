/* eslint-env jest */
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, fireEvent } from 'react-testing-library';

import Photo from '../Photo';

const comments = {
  test: [
    {
      text: 'hey!',
      user: 'kingmob'
    }
  ]
}

const post = {
  caption: 'This is a test',
  code: 'test',
  display_src: 'test.jpg',
  id: '7',
  likes: 7
};

const incrementLikes = jest.fn();

const mockedProps = {
  post,
  incrementLikes,
  comments
}

describe('<Photo/>', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <Router>
        <Photo key={8} {...mockedProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a photo correctly', () => {
    const { getByTestId } = render(
      <Router>
        <Photo key={8} {...mockedProps} />
      </Router>
    );

    expect(getByTestId('caption').textContent).toContain('This is a test');
    expect(getByTestId('count').textContent).toContain('1');
  });

  it('increments likes', () => {
    const { getByTestId } = render(
      <Router>
        <Photo key={8} {...mockedProps} />
      </Router>
    );

    fireEvent.click(getByTestId('likes'));

    expect(incrementLikes).toHaveBeenCalledTimes(1);
  });
});
