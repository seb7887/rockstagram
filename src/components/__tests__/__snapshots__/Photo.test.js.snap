// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Photo/> renders and matches snapshot 1`] = `
<DocumentFragment>
  <figure
    class="PhotoStyles__Figure-sc-1aj8yko-0 cuFNbv"
  >
    <div
      class="grid-photo-wrap"
    >
      <a
        href="/view/test"
      >
        <img
          alt="This is a test"
          class="grid-photo"
          src="test.jpg"
        />
      </a>
      <div
        class="LikeAnimation__Animation-qbj4g3-0 cogcnK"
      >
        <div>
          <span
            class="like LikeAnimation__Likes-qbj4g3-1 cgrPDV"
          >
            7
          </span>
        </div>
      </div>
    </div>
    <figcaption
      data-testid="caption"
    >
      <p>
        This is a test
      </p>
      <div
        class="control-buttons"
      >
        <button
          class="likes"
        >
          <i
            class="fa fa-heart"
            data-testid="likes"
          />
           7
        </button>
        <a
          class="button"
          href="/view/test"
        >
          <span
            class="comment-count"
            data-testid="count"
          >
            <i
              class="fa fa-comment"
            />
             1
          </span>
        </a>
      </div>
    </figcaption>
  </figure>
</DocumentFragment>
`;
